# Unit 12 - Tales from the Crypto

---


## 1. Sentiment Analysis

Use the [newsapi](https://newsapi.org/) to pull the latest news articles for Bitcoin and Ethereum and create a DataFrame of sentiment scores for each coin.

Use descriptive statistics to answer the following questions:
1. Which coin had the highest mean positive score?
2. Which coin had the highest negative score?
3. Which coin had the highest positive score?

In [1]:
# Initial imports
import warnings
warnings.filterwarnings('ignore')
import os
import pandas as pd
from dotenv import load_dotenv
import nltk as nltk
from newsapi import NewsApiClient
nltk.download('vader_lexicon')
from nltk.sentiment.vader import SentimentIntensityAnalyzer
analyzer = SentimentIntensityAnalyzer()

%matplotlib inline

[nltk_data] Downloading package vader_lexicon to /home/wh/nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!


In [2]:
load_dotenv()

True

In [3]:
# Read your api key environment variable
# YOUR CODE HERE!
key = os.getenv("news_api_key")
type(key)

str

In [4]:
# Create a newsapi client
# YOUR CODE HERE!
newsapi = NewsApiClient(api_key = key)

In [5]:
# Fetch the Bitcoin news articles
# YOUR CODE HERE!
btc_articles = newsapi.get_everything(q = 'Bitcoin', language = 'en')
len(btc_articles)

3

In [6]:
# Fetch the Ethereum news articles
# YOUR CODE HERE!
eth_articles = newsapi.get_everything(q = 'Ethereum', language = 'en')
len(eth_articles)

3

In [7]:
# Create function to get sentiments from a dictionary of articles from newsapi
def get_sentiments(dictionary):
    results = []
    for x in dictionary['articles']:
        title = x['title']
        text = x["content"]
        date = x["publishedAt"][:10]
        sentiment = analyzer.polarity_scores(text)
        compound = sentiment['compound']
        positive = sentiment['pos']
        neutral = sentiment['neu']
        negative = sentiment['neg']
        results.append({
            'Title': title,
            'Text': text,
            'Date': date,
            'Compound': compound,
            'Positive': positive,
            'Neutral': neutral,
            'Negative': negative
        })
    
    return pd.DataFrame(results)

In [8]:
# Create the Bitcoin sentiment scores DataFrame
# YOUR CODE HERE!
btc_sentiments = get_sentiments(btc_articles)
btc_sentiments.tail()

Title  \
15   What does Uber and birth control have in common?   
16  Gold is Dead? Should You Opt for Gold (Over Cr...   
17   UPDATE 1-Bitcoin rises 9.8% to $39,035 - Reuters   
18  Cryptocurrencies tumble amid China crackdown o...   
19  Boom, bust and bewildered: Bitcoin's year so f...   

                                                 Text        Date  Compound  \
15  Hello and welcome back to Equity, TechCrunchs ...  2021-06-14    0.4588   
16  This story originally appeared on MarketBeatWh...  2021-06-22    0.0772   
17  By Reuters Staff\r\n(Adds context, background)...  2021-06-13    0.0000   
18  TOKYO, June 21 (Reuters) - Cryptocurrencies tu...  2021-06-21   -0.2732   
19  A representation of virtual currency Bitcoin i...  2021-06-30    0.0000   

    Positive  Neutral  Negative  
15     0.083    0.917     0.000  
16     0.084    0.841     0.075  
17     0.000    1.000     0.000  
18     0.000    0.942     0.058  
19     0.000    1.000     0.000

In [9]:
# Create the Ethereum sentiment scores DataFrame
# YOUR CODE HERE!
eth_sentiments = get_sentiments(eth_articles)
eth_sentiments.tail()

Title  \
15   Bitcoin rises 9.8 percent to $39035 - Reuters   
16         Bitcoin falls 8.5% to $31,700 - Reuters   
17  Bitcoin rises 5.1 percent to $37,361 - Reuters   
18  Bitcoin rises 9.8 percent to $39,035 - Reuters   
19   Bitcoin rises 5.1 percent to $37361 - Reuters   

                                                 Text        Date  Compound  \
15  Bitcoin surged 9.83 % to $39,035.47 at 22:05 G...  2021-06-13   -0.2732   
16  By Reuters Staff\r\nJune 25 (Reuters) - Bitcoi...  2021-06-25   -0.3818   
17  By Reuters Staff\r\nJune 13 (Reuters) - Bitcoi...  2021-06-13    0.0000   
18  By Reuters Staff\r\nJune 13 (Reuters) - Bitcoi...  2021-06-13    0.0000   
19  Bitcoin rose 5.11 % to $37,360.63 at 18:10 GMT...  2021-06-13   -0.2732   

    Positive  Neutral  Negative  
15       0.0     0.94      0.06  
16       0.0     0.92      0.08  
17       0.0     1.00      0.00  
18       0.0     1.00      0.00  
19       0.0     0.94      0.06

In [10]:
# Describe the Bitcoin Sentiment
# YOUR CODE HERE!
btc_sentiments.describe()

Compound   Positive    Neutral   Negative
count  20.000000  20.000000  20.000000  20.000000
mean    0.155820   0.076250   0.890800   0.032900
std     0.392435   0.070288   0.084871   0.053875
min    -0.612400   0.000000   0.705000   0.000000
25%     0.000000   0.032250   0.840250   0.000000
50%     0.128000   0.065500   0.905000   0.000000
75%     0.369700   0.107250   0.955500   0.058000
max     0.840200   0.282000   1.000000   0.184000

In [11]:
# Describe the Ethereum Sentiment
# YOUR CODE HERE!
eth_sentiments.describe()

Compound   Positive    Neutral   Negative
count  20.000000  20.000000  20.000000  20.000000
mean    0.063475   0.040500   0.934350   0.025150
std     0.356132   0.058708   0.053135   0.032347
min    -0.381800   0.000000   0.806000   0.000000
25%    -0.273200   0.000000   0.920000   0.000000
50%     0.000000   0.000000   0.936500   0.000000
75%     0.361200   0.074750   0.961000   0.060250
max     0.771700   0.194000   1.000000   0.080000

### Questions:

Q: Which coin had the highest mean positive score?

A: Bitcoin: 0.076250

Q: Which coin had the highest compound score?

A: Bitcoin: 0.840200

Q. Which coin had the highest positive score?

A: Bitcoin: 0.282000

---

## 2. Natural Language Processing
---
###   Tokenizer

In this section, you will use NLTK and Python to tokenize the text for each coin. Be sure to:
1. Lowercase each word.
2. Remove Punctuation.
3. Remove Stopwords.

In [12]:
from nltk.tokenize import word_tokenize, sent_tokenize
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer, PorterStemmer
from string import punctuation
import re
nltk.download('stopwords')
nltk.download('punkt')

[nltk_data] Downloading package stopwords to /home/wh/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to /home/wh/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [13]:
# Instantiate the lemmatizer
# YOUR CODE HERE!
lemmatizer = WordNetLemmatizer()
# Create a list of stopwords
# YOUR CODE HERE!
sw = set(stopwords.words('english'))
# Expand the default stopwords list if necessary
# YOUR CODE HERE!
print(sw)

{'about', 'all', 'him', 'this', 'if', 'further', 'yourselves', 'shouldn', 'other', 'a', 'does', 'once', "don't", 'no', 'ourselves', 'her', 'needn', 'but', "you've", 'aren', 'some', 'are', 'only', 'couldn', 'same', 'myself', 'when', 'i', 'own', 'their', 'not', 'itself', 'am', 'nor', "hadn't", 'hers', 'themselves', 'from', 'while', 'won', "you'd", 'your', 're', 'off', 'hasn', 'under', 'weren', 'now', 'as', 'on', 'more', "should've", 'he', 'there', 'theirs', 'should', 'any', 'in', 'have', 'had', 'to', 'out', 'do', 'the', 'how', 'for', 'below', "couldn't", 'after', 'between', 'wasn', 'over', "she's", "shan't", 'his', "mightn't", "you'll", 'whom', "haven't", 'against', 'will', 'it', 'y', 'those', 'o', 'they', "hasn't", 'than', 'during', 't', "shouldn't", 'we', 'were', 'our', 'here', 'can', 'didn', "weren't", 'what', 'been', 'few', 'so', "you're", "needn't", 'haven', 'you', 'mightn', 'ain', 'both', 'who', 'by', 's', 'before', 'because', 'very', 'just', 'did', 'ma', 'these', 'and', "wouldn't"

In [14]:
# Complete the tokenizer function
def tokenizer(df):
    regex = re.compile("[^a-zA-Z ]")
    df['Tokens'] = ""
#     """Tokenizes text."""
#     sentence = sent_tokenize(text)['Text']
    
#     # Remove the punctuation from text
#     regex = re.compile("[^a-zA-Z]")
#     sentence = regex.sub(' ', sentence)
   
#     # Create a tokenized list of the words
#     words = word_tokenize(sentence)
    
#     # Lemmatize words into root words
#     tokens = [lemmatizer.lemmatize(x) for x in words]
   
#     # Convert the words to lowercase & Remove the stop words
#     tokens = [x.lower() for x in words if x.lower() not in sw]
    
#     return tokens
    for i in range(len(df)):
        clean = regex.sub("", df['Text'].loc[i])
        words = word_tokenize(clean)
        
        words = [x.lower() for x in words if x.lower() not in sw]
        words = [lemmatizer.lemmatize(x) for x in words]
        df['Tokens'][i] = words
    return df

In [15]:
# Create a new tokens column for Bitcoin
# YOUR CODE HERE!
btc_tokenized = tokenizer(btc_sentiments)
btc_tokenized.head()

Title  \
0  El Salvador becomes the first country to appro...   
1  El Salvador Becomes First Country to Recognize...   
2  Elon Musk says Tesla will resume Bitcoin purch...   
3  Kaseya ransomware attackers demand $70 million...   
4  Bitcoin: El Salvador makes cryptocurrency lega...   

                                                Text        Date  Compound  \
0  El Salvador's President Nayib Bukele has made ...  2021-06-09    0.8402   
1  El Salvador has become the first country in th...  2021-06-09    0.1280   
2  It's all about clean energy, it seems. \r\nElo...  2021-06-14    0.6908   
3  Filed under:\r\nThe supply chain attack has re...  2021-07-05   -0.5719   
4  image captionThe move means bitcoin will be ac...  2021-06-09    0.2732   

   Positive  Neutral  Negative  \
0     0.282    0.718     0.000   
1     0.043    0.957     0.000   
2     0.169    0.831     0.000   
3     0.111    0.705     0.184   
4     0.060    0.940     0.000   

                                              Tokens  
0  [el, salvador, president, nayib, bukele, made,...  
1  [el, salvador, become, first, country, world, ...  
2  [clean, energy, seemselon, musk, tesla, caused...  
3  [filed, underthe, supply, chain, attack, reach...  
4  [image, captionthe, move, mean, bitcoin, accep...

In [16]:
# Create a new tokens column for Ethereum
# YOUR CODE HERE!
eth_tokenized = tokenizer(eth_sentiments)
eth_tokenized.tail()

Title  \
15   Bitcoin rises 9.8 percent to $39035 - Reuters   
16         Bitcoin falls 8.5% to $31,700 - Reuters   
17  Bitcoin rises 5.1 percent to $37,361 - Reuters   
18  Bitcoin rises 9.8 percent to $39,035 - Reuters   
19   Bitcoin rises 5.1 percent to $37361 - Reuters   

                                                 Text        Date  Compound  \
15  Bitcoin surged 9.83 % to $39,035.47 at 22:05 G...  2021-06-13   -0.2732   
16  By Reuters Staff\r\nJune 25 (Reuters) - Bitcoi...  2021-06-25   -0.3818   
17  By Reuters Staff\r\nJune 13 (Reuters) - Bitcoi...  2021-06-13    0.0000   
18  By Reuters Staff\r\nJune 13 (Reuters) - Bitcoi...  2021-06-13    0.0000   
19  Bitcoin rose 5.11 % to $37,360.63 at 18:10 GMT...  2021-06-13   -0.2732   

    Positive  Neutral  Negative  \
15       0.0     0.94      0.06   
16       0.0     0.92      0.08   
17       0.0     1.00      0.00   
18       0.0     1.00      0.00   
19       0.0     0.94      0.06   

                                               Tokens  
15  [bitcoin, surged, gmt, sunday, adding, previou...  
16  [reuters, staffjune, reuters, bitcoin, dipped,...  
17  [reuters, staffjune, reuters, bitcoin, rose, g...  
18  [reuters, staffjune, reuters, bitcoin, surged,...  
19  [bitcoin, rose, gmt, sunday, adding, previous,...

---

### NGrams and Frequency Analysis

In this section you will look at the ngrams and word frequency for each coin. 

1. Use NLTK to produce the n-grams for N = 2. 
2. List the top 10 words for each coin. 

In [17]:
from collections import Counter
from nltk import ngrams

In [18]:
# Generate the Bitcoin N-grams where N=2
# YOUR CODE HERE!

In [19]:
# Generate the Ethereum N-grams where N=2
# YOUR CODE HERE!

In [20]:
# Function token_count generates the top 10 words for a given coin
def token_count(tokens, N=3):
    """Returns the top N tokens from the frequency count"""
    return Counter(tokens).most_common(N)

In [21]:
# Use token_count to get the top 10 words for Bitcoin
# YOUR CODE HERE!

In [22]:
# Use token_count to get the top 10 words for Ethereum
# YOUR CODE HERE!

---

### Word Clouds

In this section, you will generate word clouds for each coin to summarize the news for each coin

In [23]:
from wordcloud import WordCloud
import matplotlib.pyplot as plt
plt.style.use('seaborn-whitegrid')
import matplotlib as mpl
mpl.rcParams['figure.figsize'] = [20.0, 10.0]

In [24]:
# Generate the Bitcoin word cloud
# YOUR CODE HERE!

In [25]:
# Generate the Ethereum word cloud
# YOUR CODE HERE!

---
## 3. Named Entity Recognition

In this section, you will build a named entity recognition model for both Bitcoin and Ethereum, then visualize the tags using SpaCy.

In [26]:
import spacy
from spacy import displacy

In [27]:
# Download the language model for SpaCy
# !python -m spacy download en_core_web_sm

In [28]:
# Load the spaCy model
nlp = spacy.load('en_core_web_sm')

---
### Bitcoin NER

In [29]:
# Concatenate all of the Bitcoin text together
# YOUR CODE HERE!

In [30]:
# Run the NER processor on all of the text
# YOUR CODE HERE!

# Add a title to the document
# YOUR CODE HERE!

In [31]:
# Render the visualization
# YOUR CODE HERE!

In [32]:
# List all Entities
# YOUR CODE HERE!

---

### Ethereum NER

In [33]:
# Concatenate all of the Ethereum text together
# YOUR CODE HERE!

In [34]:
# Run the NER processor on all of the text
# YOUR CODE HERE!

# Add a title to the document
# YOUR CODE HERE!

In [35]:
# Render the visualization
# YOUR CODE HERE!

In [36]:
# List all Entities
# YOUR CODE HERE!

---